source/clast.c

```diff
@@ -2110,3 +2114,463 @@ void clast_filter(struct clast_stmt *node,
+    }
+    cloog_int_clear(uj_one);
+}
+
+
+/* Returns the statements in a loop body that are at the same nesting depth as
+ * the statement s. The ordering of these statements have to be preserved during
+ * unroll jam. */
+struct clast_stmt **get_statements_to_be_unrolled(struct clast_stmt *s,
+        int *num_unroll_stmts) {
+    int num, i;
+    struct clast_stmt ** unroll_stmts;
+    struct clast_stmt *curr;
+    curr = s;
+    num = 0;
+    while (curr!=NULL && CLAST_STMT_IS_A(curr, stmt_user)) {
+        num++;
+        curr = curr->next;
+    }
+    assert (num>=1);
+    curr = s;
+    unroll_stmts = (struct clast_stmt **)malloc (num * sizeof(struct clast_stmt *));
+    for (i=0; i<num; i++) {
+        unroll_stmts[i] = curr;
+        curr = curr->next;
+    }
+    *num_unroll_stmts = num;
+    return unroll_stmts;
+}
+/* Unroll jams a statement. Current supports unroll jam of a single statement as
+ * opposed to a statement block. */
+void unroll_jam_statement(struct clast_stmt *s, unsigned ufactor,
+        const char *uj_iterator, cloog_int_t stride) {
+    struct clast_user_stmt *in;
+    struct clast_stmt *curr, *prev, *last, *sub, **unroll_stmts;
+    int num_unroll_stmts, i, j;
+
+    assert(CLAST_STMT_IS_A(s, stmt_user));
+    unroll_stmts = get_statements_to_be_unrolled(s, &num_unroll_stmts);
+    last = unroll_stmts[num_unroll_stmts-1];
+    prev = last;
+    struct clast_stmt *out = last->next;
+    for (i = 1; i < ufactor; i++) {
+        /* The new unroll offset is i*loop_stride */
+        cloog_int_t uf;
+        cloog_int_init(uf);
+        cloog_int_set_si(uf, i);
+        /* Updated access is (iterator + (i*stride)). Compute i*stride first. */
+        cloog_int_mul(uf, uf,stride);
+        for (j = 0; j< num_unroll_stmts; j++) {
+            in = (struct clast_user_stmt *)unroll_stmts[j];
+            /* sub = clast_substitutions_dup(in->substitutions); */
+            sub = clast_ast_dup(in->substitutions);
+            update_iterator_in_substitution(sub, uj_iterator, uf);
+            curr = &new_clast_user_stmt(in->domain, in->statement, sub)->stmt;
+            prev->next = curr;
+            prev = curr;
+        }
+        cloog_int_clear(uf);
+    }
+    prev->next = out;
+    free(unroll_stmts);
+}
+
+/* Updates loop stride after unroll jamming. The new stride is given
+ * by unroll (factor * loop stride). */
+void update_loop_stride(struct clast_for *loop) {
+    int ufactor;
+    cloog_int_t uf;
+    ufactor = loop->ufactor;
+    cloog_int_init(uf);
+    cloog_int_set_si(uf, ufactor);
+    cloog_int_mul(loop->stride, loop->stride, uf);
+    cloog_int_clear(uf);
+}
+
+/* Unroll jams a loop body. */
+void unroll_jam_loop_body(struct clast_stmt *s, int ufactor,
+        const char *iterator, cloog_int_t stride) {
+    struct clast_for *loop;
+    struct clast_guard *g;
+    if (!s)
+        return;
+    if (CLAST_STMT_IS_A(s, stmt_user)) {
+        /* struct clast_stmt *next_stmt = s->next; */
```

Comment:
Drop commented out code - here and below.

Reply:
This was a bug. The code was commented out by mistake. It should have been there (with some modifications) in the first place. This is fixed in the next commit.